core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -374,12 +374,6 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
-    // System property spark.yarn.app.id must be set if user code ran by AM on a YARN cluster
-    if (master == "yarn" && deployMode == "cluster" && !_conf.contains("spark.yarn.app.id")) {
```

Comment:
I see, removing this will still get other error messages if trying to run cluster mode with such way, but maybe a little odd as you mentioned.